### PR TITLE
Bugfixes to optim support

### PR DESCRIPTION
**Pull request by fritzo**
This PR fixes lots of bugs in the initial optim implementation and adds support for many missing parameters.
- Support for gradients
- Support for `lower` and `upper` bounds
- Support for `fnscale` parameter

### CODE REVIEW

**Comment by fritzo** [comment]
@perrydv Could you please do a quick review of this code?
@NLMichaud This PR should cover all uses of `optim()` needed to compile your MCEM code.
@paciorek Please let me know when 0.6-5 is released and devel is unfrozen.

File: packages/nimble/inst/CppCode/predefinedNimbleLists.cpp [hunk 1]
```diff
@@ -246,12 +247,14 @@ void OptimControlNimbleList::copyFromSEXP(SEXP S_nimList_) {
     SET_STRING_ELT(S_pxData, 0, mkChar(".xData"));
     PROTECT(S_trace = findVarInFrame(GET_SLOT(S_nimList_, S_pxData),
```
Comment (perrydv) [inline packages/nimble/inst/CppCode/predefinedNimbleLists.cpp]: Per last round of rchk, I think we when there are multiple arguments to an R API function we need to ensure everything is `PROTECT`ed, with the exception of `install` calls.  So `PROTECT(GET_SLOT(S_nimList_, S_pxData))`.  This applies to all similar lines, and the corresponding `UNPROTECT` argument needs to match the number of `PROTECT`s.

**Comment by perrydv** [comment]
I made one minor comment about adding some `PROTECT`s.  Otherwise :thumbsup: .  Thanks @fritzo!

File: packages/nimble/inst/CppCode/predefinedNimbleLists.cpp (hunk 1 above)
Comment (fritzo) [inline packages/nimble/inst/CppCode/predefinedNimbleLists.cpp]: This is entirely generated code, so the actual error is in our compiler. Could you point me towards where in our compiler these lines are being generated?

Comment (fritzo) [inline packages/nimble/inst/CppCode/predefinedNimbleLists.cpp]: Ah I see in `cppDefs_nimbleList.R`. This issue is independent of this PR, so I'll fix that in a separate PR.

Comment (perrydv) [inline packages/nimble/inst/CppCode/predefinedNimbleLists.cpp]: Thanks.  I forgot I manually fixed them for 0.6-5 but they are generated.

Comment (fritzo) [inline packages/nimble/inst/CppCode/predefinedNimbleLists.cpp]: Lots of questions:
- Did you manually fix the file marked 'DO NOT EDIT BY HAND' or another file?
- Could you point me to the commit where you did this modification, so I can merge upstream into the compiler?
- Do you know how to run `rchk`? It seems very difficult.

Comment (fritzo) [inline packages/nimble/inst/CppCode/predefinedNimbleLists.cpp]: Ah I'll try to follow the pattern in bb93076b

Comment (fritzo) [inline packages/nimble/inst/CppCode/predefinedNimbleLists.cpp]: I've merged the optim work and am addressing the independent `PROTECT` issue in #406

**Pull request merged**